src/NexusMods.DataModel/Serializers/DiskStateTreeSerializer.cs

```diff
@@ -0,0 +1,114 @@
+using System.Buffers;
+using System.Runtime.InteropServices;
+using FlatSharp;
+using NexusMods.Abstractions.DiskState;
+using NexusMods.Abstractions.GameLocators;
+using NexusMods.Archives.Nx.Enums;
+using NexusMods.Archives.Nx.Utilities;
+using NexusMods.DataModel.Serializers.DiskStateTreeSchema;
+using NexusMods.Hashing.xxHash64;
+using NexusMods.MnemonicDB.Abstractions;
+using NexusMods.MnemonicDB.Storage;
+using NexusMods.Paths;
+using NexusMods.Paths.Trees.Traits;
+using Reloaded.Memory.Extensions;
+using File = NexusMods.DataModel.Serializers.DiskStateTreeSchema.File;
+
+namespace NexusMods.DataModel.Serializers;
+
+/// <summary>
+/// We want to store the disk state tree in the datom store, so we'll need to serialize it.
+/// Long term we want to redo this, as we're just shoving potentially MB of data in a value, and
+/// that's not really what MnemonicDB is designed for. It works for now and is fairly easy to fix
+/// in the future, perhaps with MnemonicDB adding support for large values.
+///
+/// Another long-term option would be to store the tree *as* a tree in the datom store, that would
+/// require more diffing logic, but would give us the ability to perform more complex audits.
+/// </summary>
+internal class DiskStateTreeSerializer : IValueSerializer<DiskStateTree>
+{
+    public int Compare(in ReadOnlySpan<byte> a, in ReadOnlySpan<byte> b)
+    {
+        return a.SequenceCompareTo(b);
+    }
+
+    public Type NativeType => typeof(DiskStateTree);
+    public Symbol UniqueId => Symbol.Intern<DiskStateTreeSerializer>();
+    
+    public DiskStateTree Read(ReadOnlySpan<byte> buffer)
+    {
+        var decompressedSize = MemoryMarshal.Read<uint>(buffer);
+        var compressedData = buffer.SliceFast(sizeof(uint));
+
+        Files files;
+
+        if (decompressedSize == buffer.Length - sizeof(uint))
+        {
+            files = Files.Serializer.Parse(compressedData.ToArray());
+        }
+        else
+        {
+            var decompressed = new byte[decompressedSize];
+            unsafe
```

Comment:
Use `GC.AllocateUninitializedArray`. The entire array gets written to, so we don't care if it contains non-zero data.